### PR TITLE
feat(37): address page, new address form, ability to view and save ad…

### LINKED ISSUE #37: As a shopper, I can login and access My Account

### PR DESCRIPTION
**Pull request by stephenrichter**
This PR adds the ability to view and add shipping addresses.

Simulator Screenshot - iPhone 15 Pro Max - 2024-06-04 at 21 48 42

### CODE REVIEW

File: lib/account/cubits/address_state.dart [hunk 1]
```diff
@@ -0,0 +1,32 @@
+import 'package:flex_storefront/checkout/models/address.dart';
+import 'package:flex_storefront/shared/bloc_helper.dart';
+
+class AddressState extends BlocState {
```
Comment (ncuillery) [inline lib/account/cubits/address_state.dart:4]: I think we should have 2 blocs/cubits:
- one for the Address list (this one already exists)
- one the the new address form

File: lib/account/pages/address_page.dart [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+  const AddressPage({super.key});
+
+  @override
+  Widget build(BuildContext context) {
+    return BlocProvider(
+      create: (context) => AddressCubit()..loadAddresses(),
+      child: Scaffold(
+        floatingActionButton: FloatingActionButton(
+          onPressed: () {
+            context.router.push(const NewAddressRoute());
```
Comment (ncuillery) [inline lib/account/pages/address_page.dart]: pushNamed?

File: lib/account/pages/address_page.dart [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+          );
+        }
+
+        return SingleChildScrollView(
+          child: Padding(
+            padding: const EdgeInsets.all(FlexSizes.appPadding),
+            child: Column(
+              children: [
+                ...state.addresses.map((a) => AddressCard(address: a)).toList(),
+              ],
```
Comment (ncuillery) [inline lib/account/pages/address_page.dart]: simply `children: state.addresses.map((a) => AddressCard(address: a)).toList()` 😎

File: lib/account/pages/new_address_page.dart [hunk 4]
```diff
@@ -0,0 +1,196 @@
...
+                                  _formKey.currentState!.fields['line2']!.value,
+                              phone:
+                                  _formKey.currentState!.fields['phone']!.value,
+                              postalCode: _formKey
+                                  .currentState!.fields['postalCode']!.value,
+                              regionIsoCode: _formKey
+                                  .currentState!.fields['region']!.value,
+                              town:
+                                  _formKey.currentState!.fields['town']!.value,
+                            );
```
Comment (ncuillery) [inline lib/account/pages/new_address_page.dart]: I think a cleaner way to handle saving data to the backend is to configure the formBuilder to return its JSON in the shape expected by the backend, so no need to build an Address model that is re-transformed into json when sent to the backend.

So we can simply do `context.read<AddressCubit>().saveAddress(_formKey.currentState)` here
No need to adjust the model definition (`id` nullable)
Even no need for the toJson method

File: lib/router.dart [hunk 5]
```diff
@@ -100,6 +102,16 @@ class AppRouter extends _$AppRouter {
...
+        AutoRoute(
+          page: AddressRoute.page,
+          path: '/address',
+          fullscreenDialog: true,
+        ),
+        AutoRoute(
+          page: NewAddressRoute.page,
+          path: '/address/new',
+          // fullscreenDialog: true,
+        ),
```
Comment (ncuillery) [inline lib/router.dart:126]: Shouldn't we AuthGard-ing them?

File: lib/account/pages/address_page.dart (hunk 3 above)
Comment (stephenrichter) [reply lib/account/pages/address_page.dart]: ah right, I had more things in this column and then pulled them out. nice catch!

File: lib/account/pages/new_address_page.dart (hunk 4 above)
Comment (stephenrichter) [reply lib/account/pages/new_address_page.dart]: The only piece that requires a bit of transformation is the `Region` and `Country`, but will make it cleaner.

**Pull request merged**